src/column.rs

```diff
@@ -1,105 +1,152 @@
             entities,
             archetypes,
-            archetype_column_offsets,
+            archetype_column_indices,
             _marker,
         }
     }
 
     /// Get this column's component from entity if it exist
     pub fn get(&self, entity: Entity) -> Result<&T, ComponentError> {
+        let meta = self
+            .entities
+            .get(entity.id as usize)
+            .map_or(true, |meta| meta.generation == entity.generation)
+            .then(|| self.entities[entity.id as usize])
+            .ok_or(NoSuchEntity)?;
+        let archetype = self
+            .archetypes
+            .get(meta.location.archetype as usize)
+            .unwrap();
+        let state = self.archetype_column_indices[meta.location.archetype as usize]
+            .ok_or_else(MissingComponent::new::<T>)?;
         unsafe {
-            let meta = self
-                .entities
-                .get(entity.id as usize)
-                .expect("no such entity");
-            let archetype = self
-                .archetypes
-                .get(meta.location.archetype as usize)
-                .unwrap();
-            let state = self.archetype_column_offsets[meta.location.archetype as usize]
-                .ok_or_else(MissingComponent::new::<T>)?;
-            let target = NonNull::new_unchecked(
-                archetype
-                    .get_base::<T>(state)
-                    .as_ptr()
-                    .add(meta.location.index as usize),
-            );
-            archetype.borrow::<T>(state);
-            let comp = target.as_ptr().cast::<T>();
-            Ok(&*comp)
+            let target = archetype
+                .get_base::<T>(state)
+                .as_ptr()
+                .add(meta.location.index as usize)
+                .cast::<T>();
+
+            Ok(&*target)
+        }
+    }
+}
+
+unsafe impl<'a, T: Component> Send for Column<'a, T> {}
+unsafe impl<'a, T: Component> Sync for Column<'a, T> {}
+
+impl<'a, T: Component> Drop for Column<'a, T> {
+    fn drop(&mut self) {
+        self.archetype_column_indices.clear();
+        for val in self.archetypes.iter() {
+            if val.has::<T>() {
+                let state = val.get_state::<T>().unwrap();
+                val.release::<T>(state);
+            }
         }
     }
 }
 
 /// Mutable column of a single component
-pub struct ColumnMut<'a, T> {
+pub struct ColumnMut<'a, T: Component> {
     entities: &'a [EntityMeta],
     archetypes: &'a [Archetype],
-    archetype_column_offsets: Vec<Option<usize>>,
+    archetype_column_indices: Vec<Option<usize>>,
     _marker: PhantomData<T>,
 }
 
 impl<'a, T: Component> ColumnMut<'a, T> {
     pub(crate) fn new(
         entities: &'a [EntityMeta],
         archetypes: &'a [Archetype],
-        archetype_column_offsets: Vec<Option<usize>>,
         _marker: PhantomData<T>,
     ) -> Self {
+        let mut archetype_column_indices: Vec<Option<usize>> = Vec::new();
+        for i in archetypes.iter() {
```

Comment:
this change made Column quite faster but now it borrows every component T, before it only borrowed from a given entity. I am not sure if this is a good thing or bad thing

Reply:
The purpose of this interface is specifically to optimize for very heavy random access, so I think this is the correct tradeoff.